From 68d5efaa10cf0689521fe05e386e3a1e11d8cb29 Mon Sep 17 00:00:00 2001
From: Chuanhong Guo <gch981213@gmail.com>
Date: Thu, 7 Apr 2022 10:16:47 +0800
Subject: [PATCH] mtd: spinand: attach/detach MTK BMT on probe/remove

Call mtk_bmt_attach() after the MTD device is registered in
spinand_probe() and mtk_bmt_detach() before it is unregistered in
spinand_remove() to enable the MediaTek bad block management table for
SPI-NAND devices.

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>
---
--- a/drivers/mtd/nand/spi/core.c
+++ b/drivers/mtd/nand/spi/core.c
@@ -19,6 +19,7 @@
 #include <linux/string.h>
 #include <linux/spi/spi.h>
 #include <linux/spi/spi-mem.h>
+#include <linux/mtd/mtk_bmt.h>
 
 int spinand_read_reg_op(struct spinand_device *spinand, u8 reg, u8 *val)
 {
@@ -1743,6 +1744,7 @@
 	if (ret)
 		return ret;
 
+	mtk_bmt_attach(mtd);
 	ret = mtd_device_register(mtd, NULL, 0);
 	if (ret)
 		goto err_spinand_cleanup;
@@ -1750,6 +1752,7 @@
 	return 0;
 
 err_spinand_cleanup:
+	mtk_bmt_detach(mtd);
 	spinand_cleanup(spinand);
 
 	return ret;
@@ -1768,6 +1771,7 @@
 	if (ret)
 		return ret;
 
+	mtk_bmt_detach(mtd);
 	spinand_cleanup(spinand);
 
 	return 0;
